In [6]:
from google.colab import drive
drive.mount('/content/gdrive')


Mounted at /content/gdrive


In [18]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score,confusion_matrix

In [7]:
df=pd.read_csv('/content/gdrive/MyDrive/news.csv')

#Getting shape and head
df.shape
df.head()

Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]

In [8]:
#getting the labels from the DataFrame
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [9]:
#Splitting the dataset into 80 20 training and testing set

x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

In [10]:
#Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [11]:
#Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

In [12]:
#Initializing a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [13]:

#Predicting on the test set and calculating accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.58%


In [23]:
#Building confusion matrix
confusion_matrix(y_test, y_pred, labels=['FAKE','REAL'])


array([[586,  52],
       [ 42, 587]])

Here we have 586 true positives, 587 true negatives, 42 false positives, and 52 false negatives.

In [15]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=50)
model.fit(tfidf_train, y_train)
y_predict = model.predict(tfidf_test)
print(accuracy_score(y_test, y_predict))

0.9171270718232044
